README.md
# ed25519-elisabeth [![Maven Central](https://img.shields.io/maven-central/v/cafe.cryptography/ed25519-elisabeth.svg?label=Maven%20Central)](https://search.maven.org/search?q=g:%22cafe.cryptography%22%20AND%20a:%22ed25519-elisabeth%22) [![Build Status](https://travis-ci.com/cryptography-cafe/ed25519-elisabeth.svg?branch=master)](https://travis-ci.com/cryptography-cafe/ed25519-elisabeth) [![Codecov](https://img.shields.io/codecov/c/gh/cryptography-cafe/ed25519-elisabeth.svg)](https://codecov.io/gh/cryptography-cafe/ed25519-elisabeth)

A pure-Java implementation of the Ed25519 signature scheme.

Requires Java 7 or higher. Requires JDK 10 or higher to build.

# Usage

## Gradle

```
implementation 'cafe.cryptography:ed25519-elisabeth:0.1.0'
```

## Apache Maven

```
<dependency>
  <groupId>cafe.cryptography</groupId>
  <artifactId>ed25519-elisabeth</artifactId>
  <version>0.1.0</version>
</dependency>
```

# Documentation

To view the public-facing API documentation, first build it:

```sh
./gradlew javadoc
```

Then open `build/docs/javadoc/index.html` in your browser.

# About

`ed25519-elisabeth` is authored by Jack Grigg.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Security
- `Ed25519ExpandedPrivateKey.sign` no longer takes a `publicKey` argument. The
  previous API allowed the caller to control how the public key was cached in
  memory, but it created an opportunity for misuse: if two signatures were
  created using different public keys, the private scalar could be recovered
  from the signatures (see [here][pubkey-2014] and [here][pubkey-2022] for
  details). We now always cache the public key ourselves to provide a safer
  signing API.

[pubkey-2014]: https://github.com/jedisct1/libsodium/issues/170
[pubkey-2022]: https://github.com/MystenLabs/ed25519-unsafe-libs

## [0.1.0] - 2020-04-13
Initial release!


src/test/java/cafe/cryptography/ed25519/Utils.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

class Utils {
    /**
     * Converts a hex string to bytes. The string may contain whitespace for
     * readability.
     *
     * @param s the hex string to be converted.
     * @return the byte[]
     */
    public static byte[] hexToBytes(String s) {
        // Strip any internal whitespace
        s = s.replaceAll(" ", "");

        // Now parse as hex
        int len = s.length();
        if (len % 2 != 0) {
            throw new IllegalArgumentException("Hex string must have an even length");
        }
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4) + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }
}


src/test/java/cafe/cryptography/ed25519/LibraryTest.java
package cafe.cryptography.ed25519;

import org.junit.Test;
import static org.junit.Assert.*;

public class LibraryTest {
    @Test
    public void testSomeLibraryMethod() {
        Library classUnderTest = new Library();
        assertTrue("someLibraryMethod should return 'true'", classUnderTest.someLibraryMethod());
    }
}


src/test/java/com/google/security/wycheproof/JsonUtil.java
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.security.wycheproof;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonParser;
import com.google.gson.stream.JsonReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.IOException;

/** Utilities for reading test vectors in JSON format */
public class JsonUtil {

  /**
   * Reads a set of test vectors from a file.
   * @param filename the name of the file, local to the directory with the
   *        the test vectors.
   * @return a JsonObject with a test
   * @throws IOException if the test vectors could not be read.
   * @throws JsonParseException if the file is not valid JSON.
   */
  public static JsonObject getTestVectors(String filename) throws 
      IOException {
    InputStream is = JsonUtil.class.getClassLoader().getResourceAsStream(filename);
    JsonReader reader = new JsonReader(new InputStreamReader(is, UTF_8));
    JsonElement elem = JsonParser.parseReader(reader);
    return elem.getAsJsonObject();
  }

  /** 
   * Converts a JsonElement into a byte array.
   * @param element a JsonElement containing an encoded byte array. 
   *        Wycheproof represents byte arrays as hexadeciamal strings.
   * @throws ClassCastException if element is not a valid string value.
   * @throws IllegalStateException - if element contains an array.
   */
  public static byte[] asByteArray(JsonElement element) {
    String hex = element.getAsString();
    return TestUtil.hexToBytes(hex);
  }
}


src/jmh/java/cafe/cryptography/ed25519/BCBench.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2020 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.security.SecureRandom;
import java.util.concurrent.TimeUnit;

import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.CryptoException;
import org.bouncycastle.crypto.Signer;
import org.bouncycastle.crypto.generators.Ed25519KeyPairGenerator;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.Ed25519KeyGenerationParameters;
import org.bouncycastle.crypto.signers.Ed25519Signer;

import org.openjdk.jmh.annotations.*;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 2, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@State(Scope.Benchmark)
public class BCBench {
    public Ed25519KeyPairGenerator keyGenerator;
    public AsymmetricKeyParameter sk;
    public AsymmetricKeyParameter vk;
    public byte[] message;
    public byte[] signature;

    @Setup
    public void prepare() throws CryptoException {
        SecureRandom r = new SecureRandom();
        this.keyGenerator = new Ed25519KeyPairGenerator();
        this.keyGenerator.init(new Ed25519KeyGenerationParameters(r));

        AsymmetricCipherKeyPair keyPair = this.keyGenerator.generateKeyPair();
        this.sk = keyPair.getPrivate();
        this.vk = keyPair.getPublic();

        this.message = new byte[64];
        r.nextBytes(this.message);

        Signer signer = new Ed25519Signer();
        signer.init(true, this.sk);
        signer.update(this.message, 0, this.message.length);
        this.signature = signer.generateSignature();
    }

    @Benchmark
    public AsymmetricCipherKeyPair keygen() {
        return this.keyGenerator.generateKeyPair();
    }

    @Benchmark
    public byte[] sign() throws CryptoException {
        Signer signer = new Ed25519Signer();
        signer.init(true, this.sk);
        signer.update(this.message, 0, this.message.length);
        return signer.generateSignature();
    }

    @Benchmark
    public boolean verify() {
        Signer signer = new Ed25519Signer();
        signer.init(false, this.vk);
        signer.update(this.message, 0, this.message.length);
        return signer.verifySignature(this.signature);
    }
}


src/jmh/java/cafe/cryptography/ed25519/Ed25519Bench.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.security.SecureRandom;
import java.util.concurrent.TimeUnit;

import org.openjdk.jmh.annotations.*;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 2, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@State(Scope.Benchmark)
public class Ed25519Bench {
    public SecureRandom r;
    public Ed25519PrivateKey sk;
    public Ed25519ExpandedPrivateKey expsk;
    public Ed25519PublicKey vk;
    public byte[] message;
    public Ed25519Signature signature;

    @Setup
    public void prepare() {
        this.r = new SecureRandom();
        this.sk = Ed25519PrivateKey.generate(this.r);
        this.expsk = this.sk.expand();
        this.vk = this.sk.derivePublic();
        this.message = new byte[64];
        r.nextBytes(this.message);
        this.signature = this.sk.expand().sign(this.message);
    }

    @Benchmark
    public Ed25519PublicKey keygen() {
        return Ed25519PrivateKey.generate(this.r).derivePublic();
    }

    @Benchmark
    public Ed25519Signature sign() {
        return this.expsk.sign(this.message);
    }

    @Benchmark
    public boolean verify() {
        return this.vk.verify(this.message, this.signature);
    }
}


src/main/java/cafe/cryptography/ed25519/Ed25519PublicKey.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import cafe.cryptography.curve25519.CompressedEdwardsY;
import cafe.cryptography.curve25519.EdwardsPoint;
import cafe.cryptography.curve25519.InvalidEncodingException;
import cafe.cryptography.curve25519.Scalar;
import org.jetbrains.annotations.NotNull;

/**
 * An Ed25519 public key.
 */
public class Ed25519PublicKey {
    private final EdwardsPoint A;
    private final CompressedEdwardsY Aenc;

    Ed25519PublicKey(EdwardsPoint A) {
        this.A = A;
        this.Aenc = A.compress();
    }

    private Ed25519PublicKey(CompressedEdwardsY Aenc) throws InvalidEncodingException {
        this.Aenc = Aenc;
        this.A = Aenc.decompress();
    }

    /**
     * Construct an Ed25519PublicKey from an array of bytes.
     *
     * @return a public key.
     * @throws InvalidEncodingException if the input is not a valid encoding.
     */
    @NotNull
    public static Ed25519PublicKey fromByteArray(@NotNull byte[] input) throws InvalidEncodingException {
        CompressedEdwardsY Aenc = new CompressedEdwardsY(input);
        return new Ed25519PublicKey(Aenc);
    }

    /**
     * Encode the public key to its compressed 32-byte form.
     *
     * @return the encoded public key.
     */
    @NotNull
    public byte[] toByteArray() {
        return this.Aenc.toByteArray();
    }

    /**
     * Verify a signature over a message with this public key.
     *
     * @return true if the signature is valid, false otherwise.
     */
    public boolean verify(@NotNull byte[] message, @NotNull Ed25519Signature signature) {
        return this.verify(message, 0, message.length, signature);
    }

    /**
     * Verify a signature over a message with this public key.
     *
     * @return true if the signature is valid, false otherwise.
     */
    public boolean verify(@NotNull byte[] message, int offset, int length, @NotNull Ed25519Signature signature) {
        // @formatter:off
        // RFC 8032, section 5.1:
        //   PH(x)   | x (i.e., the identity function)
        //   For Ed25519, dom2(f,c) is the empty string.

        // RFC 8032, section 5.1.7:
        // 2.  Compute SHA512(dom2(F, C) || R || A || PH(M)), and interpret the
        //     64-octet digest as a little-endian integer k.
        // @formatter:on
        MessageDigest h;
        try {
            h = MessageDigest.getInstance("SHA-512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        h.update(signature.R.toByteArray());
        h.update(this.Aenc.toByteArray());
        h.update(message, offset, length);
        Scalar k = Scalar.fromBytesModOrderWide(h.digest());

        // @formatter:off
        // 3.  Check the group equation [8][S]B = [8]R + [8][k]A'. It's
        //     sufficient, but not required, to instead check [S]B = R + [k]A'.
        // @formatter:on
        EdwardsPoint Aneg = this.A.negate();
        EdwardsPoint R = EdwardsPoint.vartimeDoubleScalarMultiplyBasepoint(k, Aneg, signature.S);
        return R.compress().equals(signature.R);
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Ed25519PublicKey)) {
            return false;
        }

        Ed25519PublicKey other = (Ed25519PublicKey) obj;
        return this.Aenc.equals(other.Aenc);
    }

    @Override
    public int hashCode() {
        return this.Aenc.hashCode();
    }
}


src/main/java/cafe/cryptography/ed25519/Ed25519Signature.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.util.Arrays;

import cafe.cryptography.curve25519.CompressedEdwardsY;
import cafe.cryptography.curve25519.Scalar;
import org.jetbrains.annotations.NotNull;

/**
 * An Ed25519 signature.
 */
public class Ed25519Signature {
    final CompressedEdwardsY R;
    final Scalar S;

    Ed25519Signature(CompressedEdwardsY R, Scalar S) {
        this.R = R;
        this.S = S;
    }

    /**
     * Construct an Ed25519Signature from an array of bytes.
     *
     * @return a signature.
     */
    @NotNull
    public static Ed25519Signature fromByteArray(@NotNull byte[] input) {
        if (input.length != 64) {
            throw new IllegalArgumentException("signature length is wrong");
        }

        // RFC 8032, section 5.1.7:
        // @formatter:off
        // 1. To verify a signature [...], first split the signature into two
        //    32-octet halves.  Decode the first half as a point R, and the
        //    second half as an integer S, in the range 0 <= s < L.  [...] If
        //    any of the decodings fail (including S being out of range), the
        //    signature is invalid.
        // @formatter:on
        CompressedEdwardsY R = new CompressedEdwardsY(Arrays.copyOfRange(input, 0, 32));

        // If the four most significant bits are unset, we know the scalar is
        // guaranteed to be fully reduced modulo the order of the basepoint, and
        // thus we can skip the full check.
        Scalar S;
        if ((input[63] & 240) == 0) {
            S = Scalar.fromBits(Arrays.copyOfRange(input, 32, 64));
        } else {
            S = Scalar.fromCanonicalBytes(Arrays.copyOfRange(input, 32, 64));
        }

        return new Ed25519Signature(R, S);
    }

    /**
     * Convert this signature to an array of bytes.
     *
     * @return the encoded signature.
     */
    @NotNull
    public byte[] toByteArray() {
        // RFC 8032, section 5.1.6:
        // @formatter:off
        // 6.  Form the signature of the concatenation of R (32 octets) and the
        //     little-endian encoding of S (32 octets; the three most
        //     significant bits of the final octet are always zero).
        // @formatter:on
        byte[] Rbar = this.R.toByteArray();
        byte[] Sbar = this.S.toByteArray();
        byte[] sig = new byte[Rbar.length + Sbar.length];
        System.arraycopy(Rbar, 0, sig, 0, Rbar.length);
        System.arraycopy(Sbar, 0, sig, Rbar.length, Sbar.length);
        return sig;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Ed25519Signature)) {
            return false;
        }

        Ed25519Signature other = (Ed25519Signature) obj;
        return this.R.equals(other.R) && this.S.equals(other.S);
    }

    @Override
    public int hashCode() {
        return this.R.hashCode() ^ this.S.hashCode();
    }
}


src/main/java/cafe/cryptography/ed25519/Ed25519PrivateKey.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;

import cafe.cryptography.curve25519.Scalar;
import org.jetbrains.annotations.NotNull;

/**
 * An Ed25519 private key.
 */
public class Ed25519PrivateKey {
    private final byte[] secret;

    private Ed25519PrivateKey(byte[] secret) {
        if (secret.length != 32) {
            throw new IllegalArgumentException("Invalid private key");
        }
        this.secret = Arrays.copyOf(secret, secret.length);
    }

    /**
     * Generate an Ed25519PrivateKey from a SecureRandom.
     *
     * @return the random private key.
     */
    @NotNull
    public static Ed25519PrivateKey generate(@NotNull SecureRandom random) {
        byte[] secret = new byte[32];
        random.nextBytes(secret);
        return new Ed25519PrivateKey(secret);
    }

    /**
     * Construct an Ed25519PrivateKey from an array of bytes.
     *
     * @return a private key.
     */
    @NotNull
    public static Ed25519PrivateKey fromByteArray(@NotNull byte[] secret) {
        return new Ed25519PrivateKey(secret);
    }

    /**
     * Encode the public key to its compressed 32-byte form.
     *
     * @return the encoded public key.
     */
    @NotNull
    public byte[] toByteArray() {
        return Arrays.copyOf(this.secret, this.secret.length);
    }

    /**
     * Convert this private key into its expanded form, which can be used for
     * creating signatures.
     *
     * @return the expanded private key.
     */
    @NotNull
    public Ed25519ExpandedPrivateKey expand() {
        // @formatter:off
        // RFC 8032, section 5.1.6:
        // 1.  Hash the private key, 32 octets, using SHA-512.  Let h denote the
        //     resulting digest.  Construct the secret scalar s from the first
        //     half of the digest, and the corresponding public key A, as
        //     described in the previous section.  Let prefix denote the second
        //     half of the hash digest, h[32],...,h[63].
        //
        // RFC 8032, section 5.1.5:
        // 1.  Hash the 32-byte private key using SHA-512, storing the digest in
        //     a 64-octet large buffer, denoted h.  Only the lower 32 bytes are
        //     used for generating the public key.
        MessageDigest hasher;
        try {
            hasher = MessageDigest.getInstance("SHA-512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        hasher.update(this.secret);
        byte[] h = hasher.digest();

        byte[] lower = Arrays.copyOfRange(h, 0, 32);
        byte[] upper = Arrays.copyOfRange(h, 32, 64);

        // 2.  Prune the buffer: The lowest three bits of the first octet are
        //     cleared, the highest bit of the last octet is cleared, and the
        //     second highest bit of the last octet is set.
        lower[0] &= 248;
        lower[31] &= 63;
        lower[31] |= 64;

        // 3.  Interpret the buffer as the little-endian integer, forming a
        //     secret scalar s.
        // @formatter:on
        Scalar s = Scalar.fromBits(lower);

        return new Ed25519ExpandedPrivateKey(s, upper);
    }

    /**
     * Derive the Ed25519 public key corresponding to this private key.
     *
     * @return the public key.
     */
    @NotNull
    public Ed25519PublicKey derivePublic() {
        return this.expand().derivePublic();
    }
}


src/main/java/cafe/cryptography/ed25519/Ed25519ExpandedPrivateKey.java
/*
 * This file is part of ed25519-elisabeth.
 * Copyright (c) 2019 Jack Grigg
 * See LICENSE for licensing information.
 */

package cafe.cryptography.ed25519;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import cafe.cryptography.curve25519.CompressedEdwardsY;
import cafe.cryptography.curve25519.Constants;
import cafe.cryptography.curve25519.EdwardsPoint;
import cafe.cryptography.curve25519.Scalar;
import org.jetbrains.annotations.NotNull;

/**
 * An Ed25519 expanded private key.
 */
public class Ed25519ExpandedPrivateKey {
    private final Scalar s;

    /**
     * The prefix component of the expanded Ed25519 private key.
     *
     * Note that because the `final` keyword only makes the reference a constant,
     * the contents of this byte[] could in theory be mutated (via reflection, as
     * this field is private). This misunderstanding was a contributor to the
     * "final" security bug in Google's Java implementation of Ed25519 [0]. However,
     * the primary cause of that bug was their reuse of the prefix buffer to hold
     * the result of calculating S; we are protected from that failure mode by the
     * type-safe curve25519-elisabeth API.
     *
     * [0] https://github.com/cryptosubtlety/final-security-bug
     */
    private final byte[] prefix;

    /**
     * The public key corresponding to this private key.
     *
     * We store the public key inside the expanded private key so that we always use
     * the correct public key when creating signatures, while caching its
     * computation along with the other expanded components.
     *
     * Version 0.1.0 of ed25519-elisabeth required the caller to provide the public
     * key. This allowed the caller to control how the public key was cached in
     * memory, but it created an opportunity for misuse: if two signatures were
     * created using different public keys, the private scalar could be recovered
     * from the signatures [0] [1]. We now always cache the public key ourselves to
     * provide a safer signing API.
     *
     * [0] https://github.com/jedisct1/libsodium/issues/170
     * [1] https://github.com/MystenLabs/ed25519-unsafe-libs
     */
    private final Ed25519PublicKey publicKey;

    Ed25519ExpandedPrivateKey(Scalar s, byte[] prefix) {
        this.s = s;
        this.prefix = prefix;
        EdwardsPoint A = Constants.ED25519_BASEPOINT_TABLE.multiply(this.s);
        this.publicKey = new Ed25519PublicKey(A);
    }

    /**
     * Returns the Ed25519 public key corresponding to this expanded private key.
     *
     * @return the public key.
     */
    @NotNull
    public Ed25519PublicKey derivePublic() {
        return this.publicKey;
    }

    /**
     * Sign a message with this expanded private key.
     *
     * @return the signature.
     */
    @NotNull
    public Ed25519Signature sign(@NotNull byte[] message) {
        return this.sign(message, 0, message.length);
    }

    /**
     * Sign a message with this expanded private key.
     *
     * @return the signature.
     */
    @NotNull
    public Ed25519Signature sign(@NotNull byte[] message, int offset, int length) {
        // @formatter:off
        // RFC 8032, section 5.1:
        //   PH(x)   | x (i.e., the identity function)
        //   For Ed25519, dom2(f,c) is the empty string.

        // RFC 8032, section 5.1.6:
        // 2.  Compute SHA-512(dom2(F, C) || prefix || PH(M)), where M is the
        //     message to be signed.  Interpret the 64-octet digest as a little-
        //     endian integer r.
        // @formatter:on
        MessageDigest h;
        try {
            h = MessageDigest.getInstance("SHA-512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        h.update(this.prefix);
        h.update(message, offset, length);
        Scalar r = Scalar.fromBytesModOrderWide(h.digest());

        // @formatter:off
        // 3.  Compute the point [r]B.  For efficiency, do this by first
        //     reducing r modulo L, the group order of B.  Let the string R be
        //     the encoding of this point.
        // @formatter:on
        CompressedEdwardsY R = Constants.ED25519_BASEPOINT_TABLE.multiply(r).compress();

        // @formatter:off
        // 4.  Compute SHA512(dom2(F, C) || R || A || PH(M)), and interpret the
        //     64-octet digest as a little-endian integer k.
        // @formatter:on
        h.reset();
        h.update(R.toByteArray());
        h.update(this.publicKey.toByteArray());
        h.update(message, offset, length);
        Scalar k = Scalar.fromBytesModOrderWide(h.digest());

        // @formatter:off
        // 5.  Compute S = (r + k * s) mod L.  For efficiency, again reduce k
        //     modulo L first.
        // @formatter:on
        Scalar S = r.add(k.multiply(this.s));

        return new Ed25519Signature(R, S);
    }
}
